File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private int count = 0;
    private int result = -1;
    
    public int kthSmallest(TreeNode root, int k) {
        count = 0;
        result = -1;
        inOrderTraversal(root, k);
        return result;
    }
    
    private void inOrderTraversal(TreeNode node, int k) {
        if (node == null || count >= k) {
            return;
        }
        
        inOrderTraversal(node.left, k);
        count++;
        if (count == k) {
            result = node.val;
            return;
        }
        inOrderTraversal(node.right, k);
    }
}
    

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# class Solution:
#     def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
# # #         O(n) approach : recursive
# #         mystack=[];
# #         def dfs(curr):
            
# #             if not curr:
# #                 return ;
# #             dfs(curr.left);
# #             mystack.append(curr.val);
# #             dfs(curr.right);
# #             if mystack:
# #                 for i in range(len(mystack)):
# #                     if i== k-1:
# #                         return mystack[i];
# #         return dfs(root);

# # faster way as it is Iterative DFS method: O(n) space
#             mystack =[];
#             curr= root;
#             while mystack or curr :
#                 while curr:
#                     mystack.append(curr);
#                     curr = curr.left;
#                 curr= mystack.pop();
#                 k=k-1;
#                 if k==0:
#                     return curr.val;
#                 curr=curr.right;
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        mystack = [];
        def dfs(node):
            if not node: return ;
            dfs(node.left)
            mystack.append(node.val); 
            if len(mystack) == k:
                return;
            dfs(node.right)
        dfs(root);
        return mystack[k-1];                
                    
        